core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,187 @@
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(startedInstanceMetadataWithServicesNotRunning,
+                        stack.getStack());
+                LOGGER.info("Collected {} downscale recovery candidates for hostGroup: {}", instancesWithServicesNotRunning.size(), request.getHostGroupName());
+            }
+
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    //CHECKSTYLE:OFF
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses,
+            Optional<String> runtimeVersion) {
+        boolean servicesHealthCheckAllowed = CMRepositoryVersionUtil.isCmServicesHealthCheckAllowed(runtimeVersion);
+        Predicate<InstanceMetadataView> instanceHealthCheck =  im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            boolean hostUnhealthy = detailedHostStatuses.isHostUnHealthy(hostName);
+            boolean maintenanceMode = detailedHostStatuses.isHostInMaintenanceMode(hostName);
+            boolean hostDecommissioned = detailedHostStatuses.isHostDecommissioned(hostName);
+            boolean servicesNotRunning = detailedHostStatuses.areServicesNotRunning(hostName);
+            if (servicesHealthCheckAllowed) {
+                return (!hostUnhealthy && maintenanceMode && hostDecommissioned && servicesNotRunning)
+                        || (!hostUnhealthy && !maintenanceMode && hostDecommissioned && servicesNotRunning);
+            } else {
+                return (!hostUnhealthy && maintenanceMode && hostDecommissioned)
```

Comment:
maintainanceMode maybe unnecessary here, because both value are permitted but the remaining booleans are the same

Reply:
The conditions are very explicit there to mirror the scenarios that we want to handle, which are detailed separately in a document. Yes, it is an unnecessary check - however it is far more maintainable going forward, once more scenarios are added to this.